core/src/modules/ops/record.ts

```diff
@@ -293,159 +293,162 @@ export namespace RecordOps {
+            unique: property.unique,
+          } as typeof bulkCreates[number];
+
+          const matchIdx = bulkCreates.findIndex(
+            (bc) =>
+              bc.recordId === record.id &&
+              bc.propertyId === property.id &&
+              bc.position === position
+          );
 
-      if (bulkCreates.length > 0) {
-        try {
-          await RecordProperty.bulkCreate(bulkCreates, {
-            updateOnDuplicate: [
-              "state",
-              "unique",
-              "stateChangedAt",
-              "confirmedAt",
-              "valueChangedAt",
-              "startedAt",
-              "rawValue",
-              "invalidValue",
-              "invalidReason",
-              "updatedAt",
-            ],
-          });
-        } catch (error) {
-          // we can resque an attempted single-write due to uniqueConstraint problems
-          const attemptedRecordProperty = bulkCreates[0];
-          if (
-            bulkCreates.length === 1 &&
-            attemptedRecordProperty.unique &&
-            error instanceof UniqueConstraintError
-          ) {
-            return CLS.afterCommit(() =>
-              api.sequelize.transaction((transaction: Transaction) =>
-                RecordProperty.update(
-                  {
-                    state: "ready",
-                    rawValue: null,
-                    invalidValue: `${attemptedRecordProperty.rawValue}`,
-                    invalidReason: InvalidReasons.Duplicate,
-                    stateChangedAt: new Date(),
-                    confirmedAt: new Date(),
-                    startedAt: null,
-                  },
-                  {
-                    transaction,
-                    where: {
-                      propertyId: attemptedRecordProperty.propertyId,
-                      recordId: attemptedRecordProperty.recordId,
-                      state: "pending",
-                    },
-                  }
-                )
-              )
-            );
+          if (matchIdx >= 0) {
+            bulkCreates.splice(matchIdx, 1, bulkCreate);
           } else {
-            throw error;
+            bulkCreates.push(bulkCreate);
           }
+
+          position++;
         }
+
+        // delete old properties we didn't update
+        existingRecordProperties
+          .filter(
+            (p) =>
+              p.recordId === record.id &&
+              p.propertyId === property.id &&
+              p.position >= position
+          )
+          .map((p) => bulkDeletes.where.id.push(p.id));
       }
-      if (bulkDeletes.where.id.length > 0) {
-        await RecordProperty.destroy(bulkDeletes);
+
+      recordOffset++;
+    }
+
+    // TODO: This select may be slower than allowing the upsert to throw like we used to
+    // However, the previous upsert method was really hard to deal with because it would throw and ruin/end the transaction
+    const duplicateRecordAnds = bulkCreates
```

Comment:
This section of the method adds an extra SQL select to find all the record properties that are unique and have the same rawValues.  Previously, we were relying on `updateOnDuplicate` to handle this automatically, but that had the side-effect of ending the SQL transaction, which makes doing anything after this method not possible in a CLSAction or CLSTask.  

This new query might be slow, so we will need to test it